libs/ebpfnetsh/programs.cpp

```diff
@@ -306,28 +212,80 @@ handle_ebpf_delete_program(
         return status;
     }
 
-    // TODO: If the program is pinned, unpin the specified program.
-    // TODO (issue #83): call ebpf_program_unpin() once it exists.
+    // TODO(issue #190): If the program is pinned, unpin the specified program.
     // The temporary API ebpf_api_unpin_object requires knowing the name a priori
     // and we have no way to get it yet.
 
     // Remove from our list of programs to release our own reference if we took one.
     // If there are no other references to the program, it will be unloaded.
-    auto found = std::find_if(_programs.begin(), _programs.end(), [filename, section](program_state_t& program) {
-        return program.program_filename == filename && program.program_section == section;
-    });
-    if (found != _programs.end()) {
-        found->clean();
-        _programs.erase(found);
-    } else {
-        std::cout << "Program not found\n";
-        return ERROR_SUPPRESS_OUTPUT;
+    bpf_object* object;
+    bpf_object* next_object;
+    bpf_object__for_each_safe(object, next_object)
+    {
+        bpf_program* program;
+        bpf_object__for_each_program(program, object)
+        {
+            int program_fd = bpf_program__fd(program);
+            struct bpf_prog_info info;
+            uint32_t info_size = sizeof(info);
+            if (bpf_obj_get_info_by_fd(program_fd, &info, &info_size) < 0) {
+                continue;
+            }
+            if (info.id == id) {
+                bpf_object__close(object);
+
+                // TODO: see if the program is still loaded, in which case some other process holds
+                // a reference. Get the PID of that process and display it.
+
+                return ERROR_OKAY;
+            }
+        }
+    }
+
+    std::cout << "Program not found\n";
+    return ERROR_SUPPRESS_OUTPUT;
+}
+
+ebpf_result_t
+ebpf_program_attach_by_id(ebpf_id_t program_id, ebpf_attach_type_t attach_type)
+{
+    fd_t program_fd = bpf_prog_get_fd_by_id(program_id);
+    if (program_fd < 0) {
+        return EBPF_INVALID_ARGUMENT;
     }
 
-    // TODO: see if the program is still loaded, in which case some other process holds
-    // a reference. Get the PID of that process and display it.
+    struct bpf_link* link;
+    ebpf_result_t result = ebpf_program_attach_by_fd(program_fd, &attach_type, nullptr, 0, &link);
 
-    return ERROR_SUCCESS;
+    ebpf_close_fd(program_fd); // TODO(issue #287): _close
```

Comment:
We are not calling `ebpf_link_close()` on link object.

Reply:
done